### DISCUSSION TITLE
Wrong kWh reading from TuyaMCU

### DISCUSSION THREAD

**Discussion by tesnaa** [arendst/Tasmota#12272] [Q&A] [answered]
So I have [Zmai90](https://templates.blakadder.com/ZMAi-90.html) and it is TuyaMCU device. The device has built in kwh counters and its proving to be very accurate and I would like to use this one instead (compared to meter provided by the electricity company its 99.6% accurate). So I set the relevant dpid and function and use the SetOption72 1 command.

I think this is related to #12114  regarding converting hex values to decimals

TuyaMCU mappings
```21:09:48.427 CMD: Group 0, Index 1, Command "TUYAMCU", Data ""
21:09:48.436 MQT: stat/tasmota_936F2D/RESULT = {"TuyaMCU":[{"fnId":11,"dpId":1},{"fnId":32,"dpId":18},{"fnId":31,"dpId":19},{"fnId":33,"dpId":20},{"fnId":37,"dpId":17}]}
```

console logs:

```
20:52:57.412 MQT: tele/tasmota_936F2D/RESULT = {"TuyaReceived":{"Data":"55AA03070008110200040031DED60D","Cmnd":7,"CmndData":"110200040031DED6","DpType2Id17":3268310,"17":{"DpId":17,"DpIdType":2,"DpIdData":"0031DED6"}}}
20:52:57.421 TYA: fnId=37 is set for dpId=17
20:52:57.423 TYA: RX value 57046 from dpId 17 
20:52:57.426 TYA: Rx ID=17 Total_Power=57046
```
Like previous case the code only reads lower 2 bytes

the value should be 31DED6 = 3268310 but the code only reads DED6 = 57046

perhaps @barbudor can help to make some patch (again) :D

**Comment by barbudor** [comment, marked as answer]
I already forgot what I did 2 weeks ago 🤣
Why is this different and not already handled by my patch ?

**Reply by barbudor** [reply]
Ah ok, I changed on active_power and now it's total_power.
Let me look

**Reply by barbudor** [reply]
@tesnaa Can you test this build ?
[tasmota_tuya_32bits.zip](https://github.com/arendst/Tasmota/files/6587440/tasmota_tuya_32bits.zip)

**Reply by tesnaa** [reply]
hi @barbudor , thanks again for your help :). I tried this on my another zmai-90 (backup) since I'm away at the office dont want to risk it to main device :) 

anyway the code reads correct value, however the one displayed in the web page and reported in mqtt needs to be divided by 10

```
08:45:50.504 TYA: Send "55aa0008000007"
08:45:50.525 MQT: tele/tasmota_903CB8/RESULT = {"TuyaReceived":{"Data":"55AA03070005010100010112","Cmnd":7,"CmndData":"0101000101","DpType1Id1":1,"1":{"DpId":1,"DpIdType":1,"DpIdData":"01"}}}
08:45:50.533 TYA: fnId=11 is set for dpId=1
08:45:50.535 TYA: RX Relay-1 --> MCU State: On Current State:On
08:45:50.547 MQT: tele/tasmota_903CB8/RESULT = {"TuyaReceived":{"Data":"55AA0307000811020004000662D868","Cmnd":7,"CmndData":"11020004000662D8","DpType2Id17":418520,"17":{"DpId":17,"DpIdType":2,"DpIdData":"000662D8"}}}
08:45:50.555 TYA: fnId=37 is set for dpId=17
08:45:50.558 TYA: RX value 418520 from dpId 17 
08:45:50.560 TYA: Rx ID=17 Total_Power=418520
08:45:50.565 SRC: Backlog
08:45:50.568 CMD: Grp 0, Cmnd 'RULETIMER', Idx 1, Len 2, Data '10'
08:45:50.575 MQT: stat/tasmota_903CB8/RESULT = {"T1":10,"T2":0,"T3":0,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}
08:45:50.586 MQT: tele/tasmota_903CB8/RESULT = {"TuyaReceived":{"Data":"55AA03070008120200040000044C79","Cmnd":7,"CmndData":"120200040000044C","DpType2Id18":1100,"18":{"DpId":18,"DpIdType":2,"DpIdData":"0000044C"}}}
08:45:50.593 TYA: fnId=32 is set for dpId=18
08:45:50.596 TYA: RX value 1100 from dpId 18 
08:45:50.598 TYA: Rx ID=18 Current=1100
08:45:50.603 MQT: tele/tasmota_903CB8/RESULT = {"TuyaReceived":{"Data":"55AA03070008130200040000088CBE","Cmnd":7,"CmndData":"130200040000088C","DpType2Id19":2188,"19":{"DpId":19,"DpIdType":2,"DpIdData":"0000088C"}}}
08:45:50.611 TYA: fnId=31 is set for dpId=19
08:45:50.613 TYA: RX value 2188 from dpId 19 
08:45:50.615 TYA: Rx ID=19 Active_Power=2188
08:45:50.620 MQT: tele/tasmota_903CB8/RESULT = {"TuyaReceived":{"Data":"55AA0307000814020004000008F82B","Cmnd":7,"CmndData":"14020004000008F8","DpType2Id20":2296,"20":{"DpId":20,"DpIdType":2,"DpIdData":"000008F8"}}}
08:45:50.628 TYA: fnId=33 is set for dpId=20
08:45:50.630 TYA: RX value 2296 from dpId 20 
08:45:50.632 TYA: Rx ID=20 Voltage=2296
08:45:51.346 TYA: Send "55aa00000000ff"
08:45:51.362 MQT: tele/tasmota_903CB8/RESULT = {"TuyaReceived":{"Data":"55AA030000010104","Cmnd":0,"CmndData":"01"}}
08:45:51.372 TYA: Heartbeat
08:45:52.690 MQT: tele/tasmota_903CB8/RESULT = {"TuyaReceived":{"Data":"55AA0307000814020004000008F92C","Cmnd":7,"CmndData":"14020004000008F9","DpType2Id20":2297,"20":{"DpId":20,"DpIdType":2,"DpIdData":"000008F9"}}}
08:45:52.698 TYA: fnId=33 is set for dpId=20
08:45:52.700 TYA: RX value 2297 from dpId 20 
08:45:52.702 TYA: Rx ID=20 Voltage=2297
08:45:57.305 MQT: tele/tasmota_903CB8/STATE = {"Time":"2021-06-03T08:45:57","Uptime":"0T00:12:38","UptimeSec":758,"Heap":24,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":4,"POWER":"ON","Wifi":{"AP":1,"SSId":"th0r-IoT","BSSId":"52:D9:E7:F3:4E:17","Channel":1,"RSSI":30,"Signal":-85,"LinkCount":1,"Downtime":"0T00:00:12"}}
08:45:57.317 MQT: tele/tasmota_903CB8/SENSOR = {"Time":"2021-06-03T08:45:57","ENERGY":{"TotalStartTime":"2021-06-03T00:00:00","Total":4185.201,"Yesterday":12.832,"Today":17.603,"Period":42949672.0,"Power":218.8,"ApparentPower":252.7,"ReactivePower":126.4,"Factor":0.87,"Voltage":229.7,"Current":1.100}}
```

The voltage, power, already shows the correct value in the mqtt and the web page, only the kWh remaining :) the kWh total displayed is 4185.201 while the correct one should be 418.520

**Reply by barbudor** [reply]
That's more of a problem
I can't just divide by 10 without considering existing deployed devices.
I don't have any Tuya power devices to check interop.
Introducing a new command to configure the ratio could be a solution but would increase code size and flash usage for a feature which coverage is not known outside of this product. 
A cheap and quick hack would be a compile-time #define but that means you would systematically have to build your own binary for every release/update.

Still strange that the device is in the Template repo but no-one else has reported any of the 2 issues before.
I need to investigate little more before I suggest a change.

**Reply by barbudor** [reply]
Hi @NorthernMan54
You seem to be using a ZMAI-90 TuyaMCU power meter with Tasmota 
Would you mind joining this discussion and sharing your thoughts ?
@tesnaa has identified 2 issues with Tasmota not properly readings active_power and total_power values correctly (pickup only 2 bytes from the 4 bytes value) that I already fixed.
However it seems that his device reports total_power with a 10 scale factor.
Do you have  similar issues with your units ?
Thanks

**Reply by tesnaa** [reply]
hi @barbudor 

Upon further checking, since I actually have 3 Zmai-90, with the the 3rd is the newest one and I just able to successfully convert it to tasmota after replacing the WB3S module to ESP12E. (The other 2 I dont need to replace the module since it's still using TYWE3S) 

From the log this version to obtain the correct kWh figure it needs to be divided the value by 100 (which the tasmota correctly display it in mqtt and web page).

```
16:03:20.757 MQT: tele/tasmota_AA64C3/RESULT = {"TuyaReceived":{"Data":"55AA030700080102000400003FC51C","Cmnd":7,"CmndData":"0102000400003FC5","DpType2Id1":16325,"1":{"DpId":1,"DpIdType":2,"DpIdData":"00003FC5"}}}
16:03:20.760 TYA: fnId=37 is set for dpId=1
16:03:20.763 TYA: RX value 16325 from dpId 1 
16:03:20.766 TYA: Rx ID=1 Total_Power=16325
16:03:26.499 MQT: tele/tasmota_AA64C3/STATE = {"Time":"2021-06-03T16:03:26","Uptime":"0T20:27:26","UptimeSec":73646,"Heap":25,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":50,"MqttCount":1,"POWER":"ON","Wifi":{"AP":1,"SSId":"th0r-IoT","BSSId":"F6:92:BF:B1:36:9D","Channel":1,"RSSI":46,"Signal":-77,"LinkCount":1,"Downtime":"0T00:00:04"}}
16:03:26.510 MQT: tele/tasmota_AA64C3/SENSOR = {"Time":"2021-06-03T16:03:26","ENERGY":{"TotalStartTime":"2021-06-02T00:00:00","Total":163.250,"Yesterday":0.120,"Today":0.010,"Period":0.0,"Power":0.0,"ApparentPower":0.0,"ReactivePower":0.0,"Factor":0.00,"Voltage":220.2,"Current":0.000}}
```
Whereas the old version I noticed the value needs to be divided by 1000 to convert it to kWh. I guess you can ignore further patch (original patch I guess still works ok for newer revisions of Zmai-90 :) 

Perhaps I need to use another fnid somehow. But the documentation even does not show fnid 37 https://tasmota.github.io/docs/TuyaMCU/#fnid

So far as temporary fix I put this in HA YAML file to parse the mqtt command for older Zmai-90 revision so I can still display the kwh figures in the hassio :) Thanks again for your help.

```  
- platform: mqtt
    name: smart_meter_kwh
    state_topic: "tele/tasmota_936F2D/RESULT"
    value_template: "{{ (value_json.TuyaReceived.DpType2Id17|float) / 1000 }}"
    unit_of_measurement: 'kWh'
```

**Reply by barbudor** [reply]
Do you mean you are happy with the test build as it is and can handle everything outside of Tasmota?

**Reply by tesnaa** [reply]
yes I am happy with the test build :) thank you

**Reply by barbudor** [reply]
OK. I'll PR tonight

**Reply by NorthernMan54** [reply]
PS my config for these is here - https://github.com/NorthernMan54/homebridge-tasmota/blob/master/DEVICES.md#zmai-90-current-sensor-switch

Personally as the ability to display values is limited in HomeKit, I’m only looking at volts, current and power and never looked at total.

**Reply by barbudor** [reply]
@tesnaa PR https://github.com/arendst/Tasmota/pull/12282 merged. 
Should now be in the dev binaries
